src/spline_agent/context.py

```diff
@@ -18,66 +18,66 @@
 
 from spline_agent.datasources import DataSource
 from spline_agent.enums import WriteMode
-from spline_agent.exceptions import LineageTrackingContextNotInitialized
+from spline_agent.exceptions import LineageTrackingContextNotInitializedError
 from spline_agent.lineage_model import NameAndVersion
 
 logger = logging.getLogger(__name__)
 
 
 class LineageTrackingContext:
     def __init__(self):
-        self._name: Optional[str] = None
-        self._ins: List[DataSource] = []
-        self._out: Optional[DataSource] = None
-        self._write_mode: Optional[WriteMode] = None
-        self._system_info: Optional[NameAndVersion] = None
+        self.__name: Optional[str] = None
+        self.__ins: List[DataSource] = []
+        self.__out: Optional[DataSource] = None
+        self.__write_mode: Optional[WriteMode] = None
+        self.__system_info: Optional[NameAndVersion] = None
 
     @property
     def name(self) -> Optional[str]:
-        return self._name
+        return self.__name
 
     @name.setter
     def name(self, value: str):
         assert value is not None
-        if self._name is not None:
+        if self.__name is not None:
             logger.warning(f"Tracking context property 'name' is reassigned: "
-                           f"old value '{self._name}', new value '{value}'")
-        self._name = value
+                           f"old value '{self.__name}', new value '{value}'")
+        self.__name = value
 
     @property
     def inputs(self) -> tuple[DataSource, ...]:
-        return tuple(self._ins)
```

Comment:
`__inst` is a list, `inputs` is a tuple, which is a bit inconsistent. Perhaps there is a good reason for it, but you are basically representing 1 thing with two different data structures

Reply:
Both are different things. A readonly property `input` is a part of the class contract, while `__ins` is a private mutable member holding the state. The rational is encapsulation of the mutable state. `inputs` property doesn't return a pointer to the `__ins` as a different type. Instead, it returns a new instance of type tuple that effectively is a immutable shallow copy of the `__ins`.